### ISSUE TITLE
x/tools/gopls: don't run analyzers for std and 3rd-party packages

### ISSUE THREAD

**Issue by ShoshinNikita** [golang/go#45607]
Warning in `/src/os/exec_unix.go`:

![](https://user-images.githubusercontent.com/30220965/115107505-e95a5980-9f73-11eb-8ade-a7b9b8a26e81.png)

Warning in `github.com/gorilla/mux@v1.8.0/mux.go`:

![](https://user-images.githubusercontent.com/30220965/115107506-ef503a80-9f73-11eb-9a7f-95b0a466ac7a.png)

I think these warnings are not useful, but just distracting. It makes sense to run analyzers only for workspace packages

**Comment by findleyr** [comment]
I'm not sure there's a clear answer here, though I agree that in the example above I wouldn't want to see analysis for gorilla/mux. 
 Perhaps we should disable analysis for anything in the mod cache.
 
I think this becomes more clear-cut once multi-module workspaces are the norm. At that point we'll know which modules the user is intending to work on.

CC @stamblerre @heschi

**Comment by nezorflame** [comment]
Maybe make it optional through the `gopls` setting?

**Comment by nezorflame** [comment]
True, but we also have:
- vendor
- local replace

Both of these are usually RW, thus actionable (maybe not every time and not in the case of /vendor, but it's possible).

**Comment by findleyr** [comment]
Indeed, though I would argue we also shouldn't run analyzers for vendored packages.

To be more precise, though, we should differentiate between running an analyzer and showing the diagnostics. In some cases we may want to run analyzers on all dependencies in order to compute facts.

Perhaps the rule should be "only show analysis diagnostics for workspace packages and locally replaced packages".

**Comment by ShoshinNikita** [comment]
Should I change the title to `x/tools/gopls: don't show the diagnostics for std and 3rd-party packages`?

**Comment by findleyr** [comment]
If that would satisfy your original request, then yes :)

I've updated the title with a slight variation of that.

**Comment by findleyr** [comment]
We chatted about this request in a meeting today, and some team members like seeing all diagnostics, even if they aren't actionable. I guess, per my criteria above, this means it probably _should_ be a configuration gated feature (something like `diagnoseDependencies`).

Given that the current state is preferable for some users and mildly annoying for others, this is unlikely to be prioritized soon.

**Comment by findleyr** [comment]
See also https://github.com/golang/vscode-go/issues/1275. Perhaps the setting should be more fine grained: something like
`showDiagnostics: openPackages | workspace | all`

Where `openPackages` only shows diagnostics for packages that have open files, `workspace` shows diagnostics for all workspace packages (+local replaces), and `all` shows all diagnostics (or rather all workspace + open files in dependencies, like we do now).

Frankly there's a lot of overlap with a more general request we've had to limit the scope of certain features to open packages only, for large monorepos. @heschi has been thinking about this.
